In [1]:
import pandas as pd
import numpy as np
from geopy.geocoders import Nominatim
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [2]:
# Load data
df = pd.read_csv('earthquakes.csv')

# Display the first 5 rows of the dataframe
df.head()

id  magnitude        type  \
0  us7000necw        4.8  earthquake   
1  tx2024shcj        5.1  earthquake   
2  ci40734823        3.7  earthquake   
3  tx2024scvz        3.9  earthquake   
4  us7000ndte        4.1  earthquake   

                                        title                 date  \
0         M 4.8 - 33 km WSW of Ackerly, Texas  2024-09-17T00:49:42   
1         M 5.1 - 34 km WSW of Ackerly, Texas  2024-09-17T00:49:42   
2                M 3.7 - 6 km N of Malibu, CA  2024-09-16T11:22:08   
3  M 3.9 - 58 km S of Whites City, New Mexico  2024-09-14T17:01:06   
4  M 4.1 - 60 km S of Whites City, New Mexico  2024-09-14T17:01:06   

            time        updated  \
0  1726534182289  1726583895255   
1  1726534182183  1726672002991   
2  1726485728190  1726637414586   
3  1726333266539  1726584426218   
4  1726333266382  1726334616179   

                                                 url  \
0  https://earthquake.usgs.gov/earthquakes/eventp...   
1  https://earthquake.usgs.gov/earthquakes/eventp...   
2  https://earthquake.usgs.gov/earthquakes/eventp...   
3  https://earthquake.usgs.gov/earthquakes/eventp...   
4  https://earthquake.usgs.gov/earthquakes/eventp...   

                                           detailUrl  felt  cdi  mmi  alert  \
0  https://earthquake.usgs.gov/earthquakes/feed/v...  1893    6    5  green   
1  https://earthquake.usgs.gov/fdsnws/event/1/que...  2042    6    5  green   
2  https://earthquake.usgs.gov/fdsnws/event/1/que...  1580    4    4    NaN   
3  https://earthquake.usgs.gov/fdsnws/event/1/que...     5    3    4  green   
4  https://earthquake.usgs.gov/earthquakes/feed/v...     4    3    4  green   

     status  tsunami   sig net      code                      ids  sources  \
0  reviewed        0   994  us  7000necw  ,us7000necw,tx2024shcj,  ,us,tx,   
1  reviewed        0  1040  tx  2024shcj  ,us7000necw,tx2024shcj,  ,us,tx,   
2  reviewed        0   591  ci  40734823  ,ci40734823,us7000ne83,  ,ci,us,   
3  reviewed        0   236  tx  2024scvz  ,us7000ndte,tx2024scvz,  ,us,tx,   
4  reviewed        0   260  us  7000ndte             ,us7000ndte,     ,us,   

                                               types  nst     dmin   rms  \
0  ,dyfi,ground-failure,losspager,moment-tensor,o...   37  0.28700  0.48   
1  ,dyfi,ground-failure,losspager,moment-tensor,o...   24  0.00000  0.10   
2  ,dyfi,focal-mechanism,nearby-cities,origin,pha...  135  0.03294  0.28   
3  ,dyfi,losspager,moment-tensor,origin,phase-dat...   38  0.00000  0.10   
4        ,dyfi,losspager,origin,phase-data,shakemap,   28  0.04300  0.26   

    gap magType geometryType    depth  latitude  longitude  \
0  70.0     mww        Point   4.0440   32.3984   -102.044   
1  38.0      ml        Point   6.1279   32.4140   -102.057   
2  99.0      ml        Point  10.9000   34.0678   -118.807   
3  65.0      ml        Point   6.2024   31.6470   -104.450   
4  67.0      mb        Point   8.8140   31.6323   -104.473   

                                place  distanceKM  \
0         33 km WSW of Ackerly, Texas          33   
1         34 km WSW of Ackerly, Texas          34   
2                6 km N of Malibu, CA           6   
3  58 km S of Whites City, New Mexico          58   
4  60 km S of Whites City, New Mexico          60   

                      placeOnly                 location      continent  \
0         WSW of Ackerly, Texas           Ackerly, Texas  North America   
1         WSW of Ackerly, Texas           Ackerly, Texas  North America   
2               N of Malibu, CA               Malibu, CA  North America   
3  S of Whites City, New Mexico  Whites City, New Mexico  North America   
4  S of Whites City, New Mexico  Whites City, New Mexico  North America   

                          country subnational            city  \
0  United States of America (the)       Texas  Tarzan-Lenorah   
1  United States of America (the)       Texas  Tarzan-Lenorah   
2  United States of America (the)  California     Los Angeles 

In [3]:
# Display information about the dataframe
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1137 entries, 0 to 1136
Data columns (total 43 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               1137 non-null   object 
 1   magnitude        1137 non-null   float64
 2   type             1137 non-null   object 
 3   title            1137 non-null   object 
 4   date             1137 non-null   object 
 5   time             1137 non-null   int64  
 6   updated          1137 non-null   int64  
 7   url              1137 non-null   object 
 8   detailUrl        1137 non-null   object 
 9   felt             1137 non-null   int64  
 10  cdi              1137 non-null   int64  
 11  mmi              1137 non-null   int64  
 12  alert            764 non-null    object 
 13  status           1137 non-null   object 
 14  tsunami          1137 non-null   int64  
 15  sig              1137 non-null   int64  
 16  net              1137 non-null   object 
 17  code          

In [4]:
# Check for duplicates
df.duplicated().sum()


337

In [5]:
# Drop duplicates
df = df.drop_duplicates()

In [6]:
# Check for missing values
df.isnull().sum()

id                   0
magnitude            0
type                 0
title                0
date                 0
time                 0
updated              0
url                  0
detailUrl            0
felt                 0
cdi                  0
mmi                  0
alert              278
status               0
tsunami              0
sig                  0
net                  0
code                 0
ids                  0
sources              0
types                0
nst                  0
dmin                 0
rms                  0
gap                  0
magType              0
geometryType         0
depth                0
latitude             0
longitude            0
place                0
distanceKM           0
placeOnly            0
location             0
continent          181
country            234
subnational        305
city               344
locality             0
postcode           647
what3words           0
timezone             0
locationDetails      0
dtype: int6

## Clean country and continent

In [7]:
# Rows with missing values in country and continent
df[df['country'].isna() | df['continent'].isna()].head()


id  magnitude        type  \
19  ci40727055       3.17  earthquake   
26  ci40722367       3.14  earthquake   
66  ci40861744       3.12  earthquake   
76  us6000nexj       4.40  earthquake   
84  us7000mym7       5.20  earthquake   

                                         title                 date  \
19  M 3.2 - 22 km W of Rancho Palos Verdes, CA  2024-09-07T06:03:42   
26  M 3.1 - 21 km W of Rancho Palos Verdes, CA  2024-09-02T11:34:26   
66             M 3.1 - 22 km SSE of Malibu, CA  2024-08-04T14:10:14   
76    M 4.4 - 192 km W of Port McNeill, Canada  2024-07-23T16:12:14   
84          M 5.2 - 200 km W of Tofino, Canada  2024-07-11T16:17:20   

             time        updated  \
19  1725689022350  1726165467558   
26  1725276866420  1725666675061   
66  1722780614010  1726165705648   
76  1721751134615  1724946983871   
84  1720714640281  1724093887545   

                                                  url  \
19  https://earthquake.usgs.gov/earthquakes/eventp...   
26  https://earthquake.usgs.gov/earthquakes/eventp...   
66  https://earthquake.usgs.gov/earthquakes/eventp...   
76  https://earthquake.usgs.gov/earthquakes/eventp...   
84  https://earthquake.usgs.gov/earthquakes/eventp...   

                                            detailUrl  felt  cdi  mmi  alert  \
19  https://earthquake.usgs.gov/fdsnws/event/1/que...    11    3    2    NaN   
26  https://earthquake.usgs.gov/fdsnws/event/1/que...     5    3    2    NaN   
66  https://earthquake.usgs.gov/fdsnws/event/1/que...    45    4    2    NaN   
76  https://earthquake.usgs.gov/fdsnws/event/1/que...     5    2    2  green   
84  https://earthquake.usgs.gov/fdsnws/event/1/que...     5    3    3  green   

      status  tsunami  sig net      code  \
19  reviewed        0  158  ci  40727055   
26  reviewed        0  153  ci  40722367   
66  reviewed        0  167  ci  40861744   
76  reviewed        1  299  us  6000nexj   
84  reviewed        1  417  us  7000mym7   

                                       ids         sources  \
19                 ,ci40727055,us6000nq96,         ,ci,us,   
26                 ,ci40722367,us6000np8x,         ,ci,us,   
66                 ,ci40861744,us6000ni1z,         ,ci,us,   
76                 ,at00sh33oi,us6000nexj,         ,at,us,   
84  ,at00sggvx6,us7000mym7,usauto7000mym7,  ,at,us,usauto,   

                                                types  nst    dmin   rms  \
19  ,dyfi,nearby-cities,origin,phase-data,scitech-...   94  0.1793  0.32   
26  ,dyfi,nearby-cities,origin,phase-data,scitech-...   89  0.1739  0.37   
66  ,dyfi,nearby-cities,origin,phase-data,scitech-...   88  0.1959  0.27   
76  ,dyfi,impact-link,losspager,origin,phase-data,...  109  1.0690  0.61   
84  ,dyfi,ground-failure,impact-link,internal-mome...   40  0.9950  0.99   

      gap magType geometryType  depth  latitude  longitude  \
19  143.0      ml        Point  -0.10   33.7255   -118.599   
26  138.0      ml        Point  -0.11   33.7102   -118.628   
66  151.0      ml        Point   0.80   33.8177   -118.757   
76  146.0      mb        Point  10.00   50.6220   -129.763   
84   98.0     mww        Point  10.00   48.9111   -128.445   

                                 place  distanceKM  \
19  22 km W of Rancho Palos Verdes, CA          22   
26  21 km W of Rancho Palos Verdes, CA          21   
66             22 km SSE of Malibu, CA          22   
76    192 km W of Port McNeill, Canada         192   
84          200 km W of Tofino, Canada         200   

                       placeOnly                 location continent  \
19  W of Rancho Palos Verdes, CA  Rancho Palos Verdes, CA       NaN   
26  W of Rancho Palos Verdes, CA  Rancho Palos Verdes, CA       NaN   
66             SSE of Malibu, CA               Malibu, CA       NaN   
76     W of Port McNeill, Canada     Port McNeill, Canada       NaN   
84           W of Tofino, Canada                        T       NaN   

                           country subnational city             locality  \
1

In [8]:
# Create a clean_country column initialized with existing country values
df['clean_country'] = df['country']

# For rows where country is NaN, extract the last part after comma from placeOnly
mask = df['clean_country'].isna()
df.loc[mask, 'clean_country'] = df.loc[mask, 'placeOnly'].str.split(',').str[-1].str.strip()

In [9]:
# Create clean_continent column initialized with existing continent values
df['clean_continent'] = df['continent']

In [10]:
# Check the value counts in clean_country
df['clean_country'].value_counts()

clean_country
United States of America (the)            294
Japan                                      62
Alaska                                     55
Taiwan (Province of China)                 49
Japan region                               48
Philippines                                46
Indonesia                                  44
China                                      30
Canada                                     18
Philippines (the)                          17
Chile                                      16
Afghanistan                                12
Peru                                       11
Mexico                                     11
Argentina                                  10
Iran (Islamic Republic of)                  9
Russia                                      8
Nepal                                       6
Turkiye                                     5
New Zealand                                 4
India                                       4
Azerbaijan          

In [11]:
# Correct and standardize the country values
country_mapping = {
    'United States of America (the)': 'United States',
    'Alaska': 'United States',
    'Japan region': 'Japan',
    'Philippines (the)': 'Philippines',
    'Iran (Islamic Republic of)': 'Iran',
    'Turkiye': 'Turkey',
    'Bolivia (Plurinational State of)': 'Bolivia',
    'Tanzania, the United Republic of': 'Tanzania',
    'Congo (the Democratic Republic of the)': 'Congo',
    'south of Alaska': 'United States',
    'Florida': 'United States',
    'Viet Nam': 'Vietnam',
    'Kermadec Islands region': 'New Zealand',
    'south of the Kermadec Islands': 'New Zealand',
    'Mariana Islands region': 'US Territory',
    'central Mid-Atlantic Ridge': 'International Waters',
    
}
df['clean_country'] = df['clean_country'].replace(country_mapping)


In [12]:
# Check the value counts in clean_country
df['clean_country'].value_counts()

clean_country
United States                 351
Japan                         110
Philippines                    63
Taiwan (Province of China)     49
Indonesia                      44
China                          30
Canada                         18
Chile                          16
Afghanistan                    12
Mexico                         11
Peru                           11
Argentina                      10
Iran                            9
Russia                          8
New Zealand                     6
Nepal                           6
Turkey                          5
India                           4
Azerbaijan                      4
Greece                          3
Montenegro                      3
Myanmar                         3
Nicaragua                       2
Somalia                         2
Brazil                          2
Pakistan                        2
Bolivia                         2
El Salvador                     2
Bangladesh                      2


In [13]:
df[df['clean_country'] == 'Taiwan'].head()

Empty DataFrame
Columns: [id, magnitude, type, title, date, time, updated, url, detailUrl, felt, cdi, mmi, alert, status, tsunami, sig, net, code, ids, sources, types, nst, dmin, rms, gap, magType, geometryType, depth, latitude, longitude, place, distanceKM, placeOnly, location, continent, country, subnational, city, locality, postcode, what3words, timezone, locationDetails, clean_country, clean_continent]
Index: []

In [14]:
df['clean_continent'].value_counts()

clean_continent
North America      313
Asia               256
South America       38
Europe               6
Africa               3
Insular Oceania      1
Insluar Oceania      1
Oceania              1
Name: count, dtype: int64

In [15]:
continent_mapping = {
    # North America
    'United States': 'North America',
    'Canada': 'North America',
    'Mexico': 'North America',
    'Costa Rica': 'North America',
    'Guatemala': 'North America',
    'Nicaragua': 'North America',
    'El Salvador': 'North America',
    
    # South America
    'Chile': 'South America',
    'Peru': 'South America',
    'Argentina': 'South America',
    'Brazil': 'South America',
    'Bolivia': 'South America',
    'Ecuador': 'South America',
    
    # Asia
    'Japan': 'Asia',
    'Philippines': 'Asia',
    'Taiwan (Province of China)': 'Asia',
    'Indonesia': 'Asia',
    'China': 'Asia',
    'Afghanistan': 'Asia',
    'Iran': 'Asia',
    'Nepal': 'Asia',
    'Turkey': 'Asia',
    'India': 'Asia',
    'Azerbaijan': 'Asia',
    'Myanmar': 'Asia',
    'Pakistan': 'Asia',
    'Bangladesh': 'Asia',
    'Vietnam': 'Asia',
    
    # Europe
    'Greece': 'Europe',
    'Montenegro': 'Europe',
    'Portugal': 'Europe',
    'Russia': 'Europe',
    
    # Africa
    'Somalia': 'Africa',
    'Djibouti': 'Africa',
    'Congo': 'Africa',
    'Tanzania': 'Africa',
    
    # Oceania
    'New Zealand': 'Oceania',
    'US Territory': 'Oceania',
    
    # Special cases
    'International Waters': 'International Waters'
}

# Update clean_continent based on clean_country
df['clean_continent'] = df['clean_country'].map(continent_mapping)

In [16]:
df['clean_continent'].value_counts()

clean_continent
North America           386
Asia                    344
South America            42
Europe                   15
Oceania                   7
Africa                    5
International Waters      1
Name: count, dtype: int64

In [17]:
df['clean_country'].value_counts()

clean_country
United States                 351
Japan                         110
Philippines                    63
Taiwan (Province of China)     49
Indonesia                      44
China                          30
Canada                         18
Chile                          16
Afghanistan                    12
Mexico                         11
Peru                           11
Argentina                      10
Iran                            9
Russia                          8
New Zealand                     6
Nepal                           6
Turkey                          5
India                           4
Azerbaijan                      4
Greece                          3
Montenegro                      3
Myanmar                         3
Nicaragua                       2
Somalia                         2
Brazil                          2
Pakistan                        2
Bolivia                         2
El Salvador                     2
Bangladesh                      2


In [18]:
clean_df= df[df['country'].isna()]
clean_df.head()

id  magnitude        type  \
76   us6000nexj        4.4  earthquake   
84   us7000mym7        5.2  earthquake   
85   us7000mykj        6.4  earthquake   
92   us7000mx3m        5.7  earthquake   
131  us7000ne8y        5.5  earthquake   

                                        title                 date  \
76   M 4.4 - 192 km W of Port McNeill, Canada  2024-07-23T16:12:14   
84         M 5.2 - 200 km W of Tofino, Canada  2024-07-11T16:17:20   
85       M 6.4 - 210 km WSW of Tofino, Canada  2024-07-11T15:08:47   
92       M 5.7 - 233 km WSW of Tofino, Canada  2024-07-04T16:15:08   
131        M 5.5 - 40 km E of Nichinan, Japan  2024-09-16T14:13:47   

              time        updated  \
76   1721751134615  1724946983871   
84   1720714640281  1724093887545   
85   1720710527040  1726258082875   
92   1720109708888  1726177948040   
131  1726496027151  1726585901040   

                                                   url  \
76   https://earthquake.usgs.gov/earthquakes/eventp...   
84   https://earthquake.usgs.gov/earthquakes/eventp...   
85   https://earthquake.usgs.gov/earthquakes/eventp...   
92   https://earthquake.usgs.gov/earthquakes/eventp...   
131  https://earthquake.usgs.gov/earthquakes/eventp...   

                                             detailUrl  felt  cdi  mmi  alert  \
76   https://earthquake.usgs.gov/fdsnws/event/1/que...     5    2    2  green   
84   https://earthquake.usgs.gov/fdsnws/event/1/que...     5    3    3  green   
85   https://earthquake.usgs.gov/fdsnws/event/1/que...    79    3    4  green   
92   https://earthquake.usgs.gov/fdsnws/event/1/que...    11    3    3  green   
131  https://earthquake.usgs.gov/fdsnws/event/1/que...     5    5    5  green   

       status  tsunami  sig net      code  \
76   reviewed        1  299  us  6000nexj   
84   reviewed        1  417  us  7000mym7   
85   reviewed        1  657  us  7000mykj   
92   reviewed        1  503  us  7000mx3m   
131  reviewed        0  468  us  7000ne8y   

                                                   ids               sources  \
76                             ,at00sh33oi,us6000nexj,               ,at,us,   
84              ,at00sggvx6,us7000mym7,usauto7000mym7,        ,at,us,usauto,   
85   ,usauto7000mykj,ew1720710590,us7000mykj,at00sg...  ,usauto,ew,us,at,pt,   
92   ,us7000mx3m,at00sg3x5h,usauto7000mx3m,pt24186000,     ,us,at,usauto,pt,   
131                        ,us7000ne8y,usauto7000ne8y,           ,us,usauto,   

                                                 types  nst   dmin   rms  \
76   ,dyfi,impact-link,losspager,origin,phase-data,...  109  1.069  0.61   
84   ,dyfi,ground-failure,impact-link,internal-mome...   40  0.995  0.99   
85   ,associate,dyfi,general-link,ground-failure,im...   95  0.882  0.94   
92   ,dyfi,impact-link,internal-moment-tensor,inter...   84  0.733  1.02   
131  ,dyfi,internal-moment-tensor,losspager,moment-...  101  1.154  1.27   

       gap magType geometryType   depth  latitude  longitude  \
76   146.0      mb        Point  10.000   50.6220   -129.763   
84    98.0     mww        Point  10.000   48.9111   -128.445   
85    76.0     mww        Point  10.000   48.8127   -128.725   
92    76.0     mww        Point  10.000   48.6926   -129.003   
131   57.0     mww        Point  15.711   31.6412    131.796   

                                place  distanceKM                  placeOnly  \
76   192 km W of Port McNeill, Canada         192  W of Port McNeill, Canada   
84         200 km W of Tofino, Canada         200        W of Tofino, Canada   
85       210 km WSW of Tofino, Canada         210      WSW of Tofino, Canada   
92       233 km WSW of Tofino, Canada         233      WSW of Tofino, Canada   
131        40 km E of Nichinan, Japan          40       E of Nichinan, Japan   

                 location continent country subnational city   locality  \
76   Port McNeill, Canada       NaN     NaN         NaN  NaN  Etc/GMT+9   
84                      T       NaN     NaN         NaN

In [19]:
clean_df.shape

(234, 45)

In [20]:
df.isnull().sum()

id                   0
magnitude            0
type                 0
title                0
date                 0
time                 0
updated              0
url                  0
detailUrl            0
felt                 0
cdi                  0
mmi                  0
alert              278
status               0
tsunami              0
sig                  0
net                  0
code                 0
ids                  0
sources              0
types                0
nst                  0
dmin                 0
rms                  0
gap                  0
magType              0
geometryType         0
depth                0
latitude             0
longitude            0
place                0
distanceKM           0
placeOnly            0
location             0
continent          181
country            234
subnational        305
city               344
locality             0
postcode           647
what3words           0
timezone             0
locationDetails      0
clean_count

## Clean date and time

In [21]:
# Convert date string and extract month and year
df['clean_date'] = pd.to_datetime(df['date'], format='ISO8601')
df['month'] = pd.to_datetime(df['date'], format='ISO8601').dt.month  
df['year'] = pd.to_datetime(df['date'], format='ISO8601').dt.year

In [22]:
df.head()

id  magnitude        type  \
0  us7000necw        4.8  earthquake   
1  tx2024shcj        5.1  earthquake   
2  ci40734823        3.7  earthquake   
3  tx2024scvz        3.9  earthquake   
4  us7000ndte        4.1  earthquake   

                                        title                 date  \
0         M 4.8 - 33 km WSW of Ackerly, Texas  2024-09-17T00:49:42   
1         M 5.1 - 34 km WSW of Ackerly, Texas  2024-09-17T00:49:42   
2                M 3.7 - 6 km N of Malibu, CA  2024-09-16T11:22:08   
3  M 3.9 - 58 km S of Whites City, New Mexico  2024-09-14T17:01:06   
4  M 4.1 - 60 km S of Whites City, New Mexico  2024-09-14T17:01:06   

            time        updated  \
0  1726534182289  1726583895255   
1  1726534182183  1726672002991   
2  1726485728190  1726637414586   
3  1726333266539  1726584426218   
4  1726333266382  1726334616179   

                                                 url  \
0  https://earthquake.usgs.gov/earthquakes/eventp...   
1  https://earthquake.usgs.gov/earthquakes/eventp...   
2  https://earthquake.usgs.gov/earthquakes/eventp...   
3  https://earthquake.usgs.gov/earthquakes/eventp...   
4  https://earthquake.usgs.gov/earthquakes/eventp...   

                                           detailUrl  felt  cdi  mmi  alert  \
0  https://earthquake.usgs.gov/earthquakes/feed/v...  1893    6    5  green   
1  https://earthquake.usgs.gov/fdsnws/event/1/que...  2042    6    5  green   
2  https://earthquake.usgs.gov/fdsnws/event/1/que...  1580    4    4    NaN   
3  https://earthquake.usgs.gov/fdsnws/event/1/que...     5    3    4  green   
4  https://earthquake.usgs.gov/earthquakes/feed/v...     4    3    4  green   

     status  tsunami   sig net      code                      ids  sources  \
0  reviewed        0   994  us  7000necw  ,us7000necw,tx2024shcj,  ,us,tx,   
1  reviewed        0  1040  tx  2024shcj  ,us7000necw,tx2024shcj,  ,us,tx,   
2  reviewed        0   591  ci  40734823  ,ci40734823,us7000ne83,  ,ci,us,   
3  reviewed        0   236  tx  2024scvz  ,us7000ndte,tx2024scvz,  ,us,tx,   
4  reviewed        0   260  us  7000ndte             ,us7000ndte,     ,us,   

                                               types  nst     dmin   rms  \
0  ,dyfi,ground-failure,losspager,moment-tensor,o...   37  0.28700  0.48   
1  ,dyfi,ground-failure,losspager,moment-tensor,o...   24  0.00000  0.10   
2  ,dyfi,focal-mechanism,nearby-cities,origin,pha...  135  0.03294  0.28   
3  ,dyfi,losspager,moment-tensor,origin,phase-dat...   38  0.00000  0.10   
4        ,dyfi,losspager,origin,phase-data,shakemap,   28  0.04300  0.26   

    gap magType geometryType    depth  latitude  longitude  \
0  70.0     mww        Point   4.0440   32.3984   -102.044   
1  38.0      ml        Point   6.1279   32.4140   -102.057   
2  99.0      ml        Point  10.9000   34.0678   -118.807   
3  65.0      ml        Point   6.2024   31.6470   -104.450   
4  67.0      mb        Point   8.8140   31.6323   -104.473   

                                place  distanceKM  \
0         33 km WSW of Ackerly, Texas          33   
1         34 km WSW of Ackerly, Texas          34   
2                6 km N of Malibu, CA           6   
3  58 km S of Whites City, New Mexico          58   
4  60 km S of Whites City, New Mexico          60   

                      placeOnly                 location      continent  \
0         WSW of Ackerly, Texas           Ackerly, Texas  North America   
1         WSW of Ackerly, Texas           Ackerly, Texas  North America   
2               N of Malibu, CA               Malibu, CA  North America   
3  S of Whites City, New Mexico  Whites City, New Mexico  North America   
4  S of Whites City, New Mexico  Whites City, New Mexico  North America   

                          country subnational            city  \
0  United States of America (the)       Texas  Tarzan-Lenorah   
1  United States of America (the)       Texas  Tarzan-Lenorah   
2  United States of America (the)  California     Los Angeles 

## Clean aleart

In [23]:
# Check the unique values in alert
df['alert'].unique()


array(['green', nan, 'yellow', 'red', 'orange'], dtype=object)

In [24]:
# Check the value counts in alert
df['alert'].value_counts()


alert
green     485
yellow     28
red         6
orange      3
Name: count, dtype: int64

In [25]:
# Check the distribution of alert levels
print("Alert levels distribution:")
print(df['alert'].value_counts(dropna=False))

# Check the relationship between alert and magnitude
print("\nMean magnitude for each alert level:")
print(df.groupby('alert')['magnitude'].mean())

# Check the distribution of magnitude for each alert level
print("\nMagnitude distribution by alert level:")
print(df.groupby('alert')['magnitude'].describe())

Alert levels distribution:
alert
green     485
NaN       278
yellow     28
red         6
orange      3
Name: count, dtype: int64

Mean magnitude for each alert level:
alert
green     5.381670
orange    6.300000
red       6.783333
yellow    6.203571
Name: magnitude, dtype: float64

Magnitude distribution by alert level:
        count      mean       std  min    25%  50%    75%  max
alert                                                         
green   485.0  5.381670  0.615967  3.5  5.200  5.5  5.700  7.1
orange    3.0  6.300000  0.000000  6.3  6.300  6.3  6.300  6.3
red       6.0  6.783333  0.808497  5.4  6.475  7.0  7.375  7.5
yellow   28.0  6.203571  0.747084  5.3  5.675  6.0  6.800  7.6


The relationship between magnitude and alert:
- green: mean = 5.38, range from 3.5 to 7.1
- yellow: mean = 6.20, range from 5.3 to 7.6
- orange: mean = 6.30, all values are 6.3
- red: mean = 6.78, range from 5.4 to 7.5

In [26]:
# Determine earthquake alert level based on magnitude
def determine_alert_level(row):
    """
    Determine earthquake alert level based on magnitude.
    Only determine new alert level if current alert is missing.
    """
    # Check if alert is missing
    if pd.isna(row['alert']):
        magnitude = row['magnitude']
        
        # Low impact earthquakes (< 5.0)
        if magnitude < 5.0:
            return 'green'
        
        # Light to moderate earthquakes (5.0 - 5.9)
        if magnitude >= 5.0 and magnitude < 6.0:
            if magnitude < 5.5:
                # Light earthquakes (5.0 - 5.4)
                return 'green'
            else:
                # Moderate earthquakes (5.5 - 5.9)
                return 'yellow'
        
        # Strong earthquakes (6.0 - 6.4)
        if magnitude >= 6.0 and magnitude < 6.5:
            return 'yellow'
        
        # Major earthquakes (6.5 - 6.9)
        if magnitude >= 6.5 and magnitude < 7.0:
            return 'orange'
        
        # Great earthquakes (7.0+)
        if magnitude >= 7.0:
            return 'red'
    else:
        return row['alert']

# Apply function to entire dataframe
df['alert'] = df.apply(determine_alert_level, axis=1)

In [27]:
# Verify the distribution after filling
print("New alert level distribution:")
print(df['alert'].value_counts(dropna=False))

New alert level distribution:
alert
green     762
yellow     29
red         6
orange      3
Name: count, dtype: int64


## Extract cleaned dataset

In [28]:
df.columns


Index(['id', 'magnitude', 'type', 'title', 'date', 'time', 'updated', 'url',
       'detailUrl', 'felt', 'cdi', 'mmi', 'alert', 'status', 'tsunami', 'sig',
       'net', 'code', 'ids', 'sources', 'types', 'nst', 'dmin', 'rms', 'gap',
       'magType', 'geometryType', 'depth', 'latitude', 'longitude', 'place',
       'distanceKM', 'placeOnly', 'location', 'continent', 'country',
       'subnational', 'city', 'locality', 'postcode', 'what3words', 'timezone',
       'locationDetails', 'clean_country', 'clean_continent', 'clean_date',
       'month', 'year'],
      dtype='object')

In [29]:
# Select only columns of interest
earthquake = df[['id','latitude', 'longitude', 'place', 'clean_country', 'clean_continent', 
                 'magnitude', 'depth', 'sig', 'felt', 'mmi', 'cdi', 'dmin', 'rms', 'gap', 'magType',
                 'alert', 'tsunami', 'nst',
                 'clean_date', 'month', 'year']]


In [30]:
# Change column names
earthquake.columns = ['ID', 'Latitude', 'Longitude', 'Place', 'Country', 'Continent', 
                 'Magnitude', 'Depth', 'Sig', 'Felt', 'MMI', 'CDI', 'Dmin', 'RMS', 'Gap', 'MagType',
                 'Alert', 'Tsunami', 'Nst',
                 'Date', 'Month', 'Year']

In [31]:
earthquake.head()

ID  Latitude  Longitude                               Place  \
0  us7000necw   32.3984   -102.044         33 km WSW of Ackerly, Texas   
1  tx2024shcj   32.4140   -102.057         34 km WSW of Ackerly, Texas   
2  ci40734823   34.0678   -118.807                6 km N of Malibu, CA   
3  tx2024scvz   31.6470   -104.450  58 km S of Whites City, New Mexico   
4  us7000ndte   31.6323   -104.473  60 km S of Whites City, New Mexico   

         Country      Continent  Magnitude    Depth   Sig  Felt  MMI  CDI  \
0  United States  North America        4.8   4.0440   994  1893    5    6   
1  United States  North America        5.1   6.1279  1040  2042    5    6   
2  United States  North America        3.7  10.9000   591  1580    4    4   
3  United States  North America        3.9   6.2024   236     5    4    3   
4  United States  North America        4.1   8.8140   260     4    4    3   

      Dmin   RMS   Gap MagType  Alert  Tsunami  Nst                Date  \
0  0.28700  0.48  70.0     mww  green        0   37 2024-09-17 00:49:42   
1  0.00000  0.10  38.0      ml  green        0   24 2024-09-17 00:49:42   
2  0.03294  0.28  99.0      ml  green        0  135 2024-09-16 11:22:08   
3  0.00000  0.10  65.0      ml  green        0   38 2024-09-14 17:01:06   
4  0.04300  0.26  67.0      mb  green        0   28 2024-09-14 17:01:06   

   Month  Year  
0      9  2024  
1      9  2024  
2      9  2024  
3      9  2024  
4      9  2024

In [32]:
earthquake.info()

<class 'pandas.core.frame.DataFrame'>
Index: 800 entries, 0 to 1135
Data columns (total 22 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   ID         800 non-null    object        
 1   Latitude   800 non-null    float64       
 2   Longitude  800 non-null    float64       
 3   Place      800 non-null    object        
 4   Country    800 non-null    object        
 5   Continent  800 non-null    object        
 6   Magnitude  800 non-null    float64       
 7   Depth      800 non-null    float64       
 8   Sig        800 non-null    int64         
 9   Felt       800 non-null    int64         
 10  MMI        800 non-null    int64         
 11  CDI        800 non-null    int64         
 12  Dmin       800 non-null    float64       
 13  RMS        800 non-null    float64       
 14  Gap        800 non-null    float64       
 15  MagType    800 non-null    object        
 16  Alert      800 non-null    object        
 17  T

In [33]:
earthquake.describe()


Latitude   Longitude   Magnitude       Depth          Sig  \
count  800.000000  800.000000  800.000000  800.000000   800.000000   
mean    28.072626   -8.555882    4.829937   38.377687   428.612500   
min    -43.706400 -179.807000    3.000000   -0.250000   138.000000   
25%     23.806650 -104.509250    3.800000    7.463075   234.000000   
50%     31.673300  -71.563450    5.200000   10.000000   449.000000   
75%     37.800325  123.749750    5.600000   35.000000   504.000000   
max     68.176100  179.972000    7.600000  639.503000  2419.000000   
std     19.800936  118.818660    1.036039   83.013337   255.271668   

                Felt         MMI        CDI        Dmin         RMS  \
count     800.000000  800.000000  800.00000  800.000000  800.000000   
mean      474.166250    4.348750    2.97125    1.271233    0.596302   
min         0.000000    1.000000    0.00000    0.000000    0.000000   
25%         0.000000    4.000000    0.00000    0.100000    0.355000   
50%         2.000000    4.000000    3.00000    0.619000    0.630000   
75%        27.000000    5.000000    5.00000    1.902500    0.780000   
max    183786.000000    9.000000    9.00000   12.457000    2.520000   
std      6746.868187    1.501646    2.58530    1.641513    0.304249   

              Gap     Tsunami         Nst                           Date  \
count  800.000000  800.000000  800.000000                            800   
mean    56.146038    0.068750  108.411250  2024-04-04 22:24:44.096249856   
min      0.000000    0.000000    0.000000            2023-06-23 17:39:13   
25%     30.000000    0.000000   36.000000  2023-12-16 18:23:08.750000128   
50%     49.000000    0.000000   97.000000            2024-04-22 10:50:30   
75%     69.000000    0.000000  147.000000  2024-07-30 01:28:26.750000128   
max    256.000000    1.000000  619.000000            2024-09-18 12:02:12   
std     41.026540    0.253187   87.965843                            NaN   

            Month         Year  
count  800.000000   800.000000  
mean     6.805000  2023.740000  
min      1.000000  2023.000000  
25%      4.000000  2023.000000  
50%      7.000000  2024.000000  
75%      9.000000  2024.000000  
max     12.000000  2024.000000  
std      3.081208     0.438909

In [34]:
# # Save cleaned dataset
# earthquake.to_excel('clean_earthquakes.xlsx', index=False)

In [35]:
# Check min and max dates in the dataset
print("Earliest date:", earthquake['Date'].min())
print("Latest date:", earthquake['Date'].max())


Earliest date: 2023-06-23 17:39:13
Latest date: 2024-09-18 12:02:12


In [36]:
# Count earthquakes by month
monthly_counts = earthquake['Month'].value_counts().sort_index()
print("\nNumber of earthquakes by month:")
print(monthly_counts)


Number of earthquakes by month:
Month
1      50
2      51
3      40
4      79
5      41
6      40
7     115
8     142
9      95
10     65
11     19
12     63
Name: count, dtype: int64


In [37]:
# Count earthquakes in United States
us_earthquakes = earthquake[earthquake['Country'] == 'United States']
print("\nNumber of earthquakes in United States:", len(us_earthquakes))



Number of earthquakes in United States: 351


In [38]:
# Calculate average magnitude by country and sort in descending order
avg_magnitude_by_country = earthquake.groupby('Country')['Magnitude'].mean().sort_values(ascending=False)
print("\nAverage earthquake magnitude by country:")
print(avg_magnitude_by_country)



Average earthquake magnitude by country:
Country
Brazil                        6.550000
Guatemala                     6.200000
Russia                        6.037500
Afghanistan                   5.858333
Peru                          5.845455
Argentina                     5.810000
Chile                         5.800000
Greece                        5.800000
Japan                         5.701818
Philippines                   5.700000
Taiwan (Province of China)    5.693878
Indonesia                     5.640909
Myanmar                       5.600000
Costa Rica                    5.600000
Bolivia                       5.550000
Nicaragua                     5.550000
El Salvador                   5.550000
Nepal                         5.533333
China                         5.530000
Tanzania                      5.500000
Somalia                       5.500000
International Waters          5.500000
Turkey                        5.480000
Montenegro                    5.400000
Ecuador       

In [39]:
# Calculate total felt earthquakes by country and sort in descending order
felt_by_country = earthquake.groupby('Country')['Felt'].sum().sort_values(ascending=False)
print("\nNumber of people who felt earthquakes by country:")
print(felt_by_country)



Number of people who felt earthquakes by country:
Country
United States                 361959
Portugal                        2869
Taiwan (Province of China)      2680
Japan                           1706
Canada                          1217
Nepal                           1180
Indonesia                       1158
China                           1095
Philippines                     1040
Montenegro                       744
Chile                            613
Bangladesh                       438
Argentina                        366
Azerbaijan                       362
Afghanistan                      283
Peru                             209
Pakistan                         204
Mexico                           198
Myanmar                          122
New Zealand                      119
El Salvador                       98
Nicaragua                         92
Turkey                            92
India                             91
Greece                            90
Guatemala       

In [40]:
# Count earthquakes and tsunamis by continent
tsunami_by_continent = earthquake.groupby('Continent')[['Tsunami']].sum()
earthquake_counts = earthquake['Continent'].value_counts()

# Combine into a single DataFrame
continent_summary = pd.DataFrame({
    'Total_Earthquakes': earthquake_counts,
    'Tsunamis': tsunami_by_continent['Tsunami']
}).sort_values('Total_Earthquakes', ascending=False)

print("\nEarthquakes and tsunamis by continent:")
print(continent_summary)



Earthquakes and tsunamis by continent:
                      Total_Earthquakes  Tsunamis
Continent                                        
North America                       386        34
Asia                                344        16
South America                        42         3
Europe                               15         2
Oceania                               7         0
Africa                                5         0
International Waters                  1         0
